### PR TITLE
Make PlaceMention a non-mutating use.

### LINKED ISSUE #110781: UB: edition 2015/18 with -Zmir-opt-level=2: destructing tuple inside closure
I tried this code:

```rust
fn main() {
    let t = (String::from("Hello"), String::from("World"));

    let c = ||  {
        let (_, _) = t;
    };

    c();
}
```

You need edition 2015 or 2018 to reproduce this.

`RUSTFLAGS="-Zmir-opt-level=2" ~/.cargo/bin/cargo miri run`

````
error: Undefined Behavior: using uninitialized data, but this operation requires initialized memory
 --> src/main.rs:5:22
  |
5 |         let (_, _) = t;
  |                      ^ using uninitialized data, but this operation requires initialized memory
  |
  = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
  = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
  = note: BACKTRACE:
  = note: inside closure at src/main.rs:5:22: 5:23
note: inside `main`
 --> src/main.rs:8:5
  |
8 |     c();
  |     ^^^

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error; 1 warning emitted
````

```
  = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5: 250:71
  = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:134:18: 134:21
  = note: inside closure at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:166:18: 166:82
  = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:284:13: 284:31
  = note: inside `std::panicking::r#try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:490:40: 490:43
  = note: inside `std::panicking::r#try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:454:19: 454:81
  = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:140:14: 140:33
  = note: inside closure at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:148:48: 148:73
  = note: inside `std::panicking::r#try::do_call::<[closure@std::rt::lang_start_internal::{closure#2}], isize>` at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:490:40: 490:43
  = note: inside `std::panicking::r#try::<isize, [closure@std::rt::lang_start_internal::{closure#2}]>` at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:454:19: 454:81
  = note: inside `std::panic::catch_unwind::<[closure@std::rt::lang_start_internal::{closure#2}], isize>` at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:140:14: 140:33
  = note: inside `std::rt::lang_start_internal` at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:148:20: 148:98
  = note: inside `std::rt::lang_start::<()>` at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:165:17: 170:6

error: aborting due to previous error; 1 warning emitted
```

### PR DESCRIPTION
**Pull request by cjgillot**
Fixes https://github.com/rust-lang/rust/issues/110781

r? @JakobDegen 

I don't agree with your statement in https://github.com/rust-lang/rust/issues/110781#issuecomment-1520841434. I suggest that we start fixing `PlaceContext` to be accurate enough for optimizations to use it. This structure is very convenient to use in visitors, and we perhaps have an opportunity to make it less of a footgun.

### CODE REVIEW

File: compiler/rustc_middle/src/mir/visit.rs [hunk 1]
```diff
@@ -1251,6 +1251,8 @@ pub enum NonMutatingUseContext {
     UniqueBorrow,
     /// AddressOf for *const pointer.
     AddressOf,
+    /// PlaceMention statement.
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/visit.rs:1254]: Might be worth adding a clarifying comment on why this is here, given that it used to be in the wrong category.

File: compiler/rustc_borrowck/src/type_check/mod.rs [hunk 2]
```diff
@@ -771,12 +771,10 @@ impl<'a, 'b, 'tcx> TypeVerifier<'a, 'b, 'tcx> {
...
             PlaceContext::MutatingUse(_) => ty::Invariant,
-            PlaceContext::NonUse(StorageDead | StorageLive | PlaceMention | VarDebugInfo) => {
-                ty::Invariant
-            }
+            PlaceContext::NonUse(StorageDead | StorageLive | VarDebugInfo) => ty::Invariant,
             PlaceContext::NonMutatingUse(
-                Inspect | Copy | Move | SharedBorrow | ShallowBorrow | UniqueBorrow | AddressOf
-                | Projection,
+                Inspect | Copy | Move | PlaceMention | SharedBorrow | ShallowBorrow | UniqueBorrow
+                | AddressOf | Projection,
```
Comment (JakobDegen) [inline compiler/rustc_borrowck/src/type_check/mod.rs:777]: Looks like we're going from invariant to covariant here? If intentional, probably a behavior change that would need a types FCP?

**Comment by JakobDegen** [comment]
> I suggest that we start fixing PlaceContext to be accurate enough for optimizations to use it.

I'd be very happy to do that if we can. My concern is just that `PlaceMention` is (presumably) the way it is because it makes sense elsewhere. If we start redefining it to be useful to opts, do we just move the bugs to borrowck?

File: compiler/rustc_borrowck/src/type_check/mod.rs (hunk 2 above)
Comment (cjgillot) [reply compiler/rustc_borrowck/src/type_check/mod.rs:777]: I'd consider this more of a bugfix for https://github.com/rust-lang/rust/pull/104844, which is still on nightly.

Comment (JakobDegen) [reply compiler/rustc_borrowck/src/type_check/mod.rs:777]: Ok, I'm going to cc @rust-lang/types here to get some eyes on this. From my understanding it's probably fine, but I'm definitely not a qualified reviewer

**Comment by JakobDegen** [comment]
lgtm modulo the types change

File: compiler/rustc_borrowck/src/type_check/mod.rs (hunk 2 above)
Comment (lcnr) [reply compiler/rustc_borrowck/src/type_check/mod.rs:777]: being covariant is fine here :+1:

**Comment by RalfJung** [comment]
@JakobDegen 
> lgtm modulo the variance change

@lcnr 
> being covariant is fine here +1

@bors r=JakobDegen,lcnr

**Pull request merged**